Add decorator for tracking execution statistics of check methods

djova opened the pull request:
### What does this PR do?

Adds a new decorator `tracked_method` to be used for tracking the performance of methods in a standardized way. This will significantly reduce measurement boilerplate across the DBM integrations and make the measurements less error prone. It's important to track the performance and error rates of various parts of the DBM integrations in order to better troubleshoot customer issues.

We already have a bunch of internal debug metrics tracking the execution time and error rate of various operations in DBM. For example, `dd.postgres.collect_statement_samples.time` tracks the execution time of the `_collect_statement_samples` method for postgres. All of these manual measurements will be updated to use this new measurement decorator instead.

### Motivation

Reduce boilerplate and make debugging performance measurements more standardized and less error prone.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have `changelog/` and `integration/` labels attached

justiniso commented on datadog_checks_base/datadog_checks/base/utils/db/utils.py:
```diff
@@ -278,3 +278,71 @@ def _run_job_rate_limited(self):
+    the statsd methods (i.e. histogram, count, gauge, etc). This is useful when specific tags need to be added to
+    these debug metrics in a standardized way.
+
+    :param dbms: the dbms to report as (i.e. sqlserver)
```
Can we make dbms the integration name?

djova replied:
good point, updated to use that

justiniso commented on datadog_checks_base/datadog_checks/base/utils/db/utils.py:
```diff
@@ -278,3 +278,71 @@ def _run_job_rate_limited(self):
+                result = function(self, *args, **kwargs)
+
+                elapsed_ms = (time.time() - start_time) * 1000
+                check.histogram(
```
Why can't we just use statsd instead of this logic for referencing the AgentCheck reference?

djova replied:
Discussed offline. There are several issues with using statsd (like the datadog library containing the statsd client not being currently bundled by default) so we're going to continue using the agent's built-in stats methods for now.

justiniso commented on datadog_checks_base/datadog_checks/base/utils/db/utils.py:
```diff
@@ -278,3 +278,71 @@ def _run_job_rate_limited(self):
+
+                check.log.debug("[%s %s] operation completed in %s ms", elapsed_ms)
+
+                if track_result_length and result is not None:
```
This seems like a weird metric since not everything supports `len` and also doesn't give an indication of nested sizes. Maybe `sys.getsizeof` is better?

djova replied:
Yeah but it's off by default so you'd enable it only on methods where you known you're returning a list-like thing and want to track the length of that. The example is that several of the methods we have that read out lists of results from the database would benefit from being able to track the response length in an easy way.

djova replied:
`sys.getsizeof` is not the same as the length of the list so it would give different information.

coignetp commented on datadog_checks_base/datadog_checks/base/utils/db/utils.py:
```diff
@@ -278,3 +278,71 @@ def _run_job_rate_limited(self):
+    :return: a decorator
+    """
+
+    def decorator(function):
```
If the intent is to have the tracking decorator enabled by default, there should be a config option (can be a hidden option) to disable the tracking on the user side

djova replied:
good idea, added option to disable

coignetp commented on datadog_checks_base/datadog_checks/base/utils/db/utils.py:
```diff
@@ -278,3 +278,71 @@ def _run_job_rate_limited(self):
         raise NotImplementedError()
+
+
+def dbm_tracked_method(dbms, parent_check_attr=None, track_result_length=False):
```
We can definitely use it more generally, let's move this to a new file `datadog_checks_base/datadog_checks/base/utils/tracking.py`
```suggestion
def tracked_method(integration, parent_check_attr=None, track_result_length=False):
```

coignetp reviewed (changes requested):
This looks interesting! I have a few suggestions ; did you manage to measure the overhead of this decorator?

djova commented:
> This looks interesting! I have a few suggestions ; did you manage to measure the overhead of this decorator?

Addressed all of your comments. I didn't measure the overhead. statsd is very low overhead generally and in this case the metrics are being aggregated directly in the agent via the internal API so it should be even lower overhead than typical statsd sent over the network.

coignetp reviewed (dismissed):
👍  if CI is green

ofek commented:
Just please edit the title to be more descriptive
